include/engine/guidance/route_step.hpp

```diff
@@ -71,55 +71,117 @@ struct RouteStep
+
+        return *this;
+    }
+
+    // Elongate by another step in back
+    RouteStep &ElongateBy(const RouteStep &following_step)
+    {
+        BOOST_ASSERT(geometry_end == following_step.geometry_begin + 1);
+        BOOST_ASSERT(mode == following_step.mode);
+        duration += following_step.duration;
+        distance += following_step.distance;
+
+        geometry_end = following_step.geometry_end;
+        intersections.insert(intersections.end(),
+                             following_step.intersections.begin(),
+                             following_step.intersections.end());
+
+        return *this;
+    }
+
+    // Elongate without prior knowledge of in front, or in back.
+    RouteStep &MergeWith(const RouteStep &by_step)
+    {
+        // if our own geometry ends, where the next begins, we elongate by
+        if (geometry_end == by_step.geometry_begin + 1)
+            return MakePreceededBy(by_step);
+        else
+            return ElongateBy(by_step);
+    }
+
+    // copy all strings from origin into the step, apart from rotary names
+    RouteStep &AdaptStepSignage(const RouteStep &origin)
+    {
+        name_id = origin.name_id;
+        name = origin.name;
+        pronunciation = origin.pronunciation;
+        destinations = origin.destinations;
+        ref = origin.ref;
+
+        return *this;
+    }
+
+    LaneID NumLanesToTheRight() const
     {
         return intersections.front().lanes.first_lane_from_the_right;
     }
 
-    LaneID numLanesToTheLeft() const
+    LaneID NumLanesToTheLeft() const
     {
         LaneID const total = intersections.front().lane_description.size();
         return total - (intersections.front().lanes.lanes_in_turn +
                         intersections.front().lanes.first_lane_from_the_right);
     }
 
-    auto lanesToTheLeft() const
+    auto LanesToTheLeft() const
     {
         const auto &description = intersections.front().lane_description;
-        LaneID num_lanes_left = numLanesToTheLeft();
+        LaneID num_lanes_left = NumLanesToTheLeft();
         return boost::make_iterator_range(description.begin(),
                                           description.begin() + num_lanes_left);
     }
 
-    auto lanesToTheRight() const
+    auto LanesToTheRight() const
     {
         const auto &description = intersections.front().lane_description;
-        LaneID num_lanes_right = numLanesToTheRight();
+        LaneID num_lanes_right = NumLanesToTheRight();
         return boost::make_iterator_range(description.end() - num_lanes_right, description.end());
     }
 };
 
-inline RouteStep getInvalidRouteStep()
-{
-    return {0,
-            "",
-            "",
-            "",
-            "",
-            "",
-            "",
-            0,
-            0,
-            TRAVEL_MODE_INACCESSIBLE,
-            getInvalidStepManeuver(),
-            0,
-            0,
-            {getInvalidIntersection()}};
-}
-}
-}
-}
+} // namespace guidance
```

Comment:
In general, what we could do is instead of

```c++
struct S {
  void fn()
  {
    // really long
    // and detailed impl
  }
};
```

do

```c++
struct S {
  // Does X because of Y. Return Z.
  void fn();
};

// Impl.

void S::fn()
{
  // impl.
}
```